File: python/heap/heap.py
# coding: utf-8
from typing import List


class Heap(object):
    def __init__(self, capacity: int):
        """
        :param capacity: 多大容量的一个数组
        """
        print(f"capacity:{capacity}")
        self.data: List[int] = []

    @property
    def size(self) -> int:
        return len(self.data)

    def is_empty(self) -> bool:
        return self.size == 0

    @staticmethod
    def _parent(index: int) -> int:
        """
        返回index对应的parent index

        由于索引为0的也使用，这里的计算公式发生了改变，同理试用于获取left,right child.
        :param index:
        :return:
        """
        if index == 0:
            raise ValueError("index 0 does't have parent")
        return (index - 1) // 2

    @staticmethod
    def _left_child(index: int) -> int:
        """
        返回index对应的left child index.
        :param index:
        :return:
        """
        return index * 2 + 1

    @staticmethod
    def _right_child(index: int) -> int:
        """
        返回index对应的right child index.
        :param index:
        :return:
        """
        return (index + 1) * 2

    def add(self, e: int):
        """
        添加元素
        :param e:
        :return:
        """
        self.data.append(e)
        self.shift_up(len(self.data) - 1)

    def shift_up(self, index: int):
        while index > 0 and self.data[self._parent(index)] < self.data[index]:
            # swap index and index parent data.
            self.data[index], self.data[self._parent(index)] = self.data[self._parent(index)], self.data[index]
            index = self._parent(index)

    def extract_max(self):
        """
        获取最大值
        :return:
        """
        try:
            e = self.data[0]
        except IndexError:
            raise IndexError('empty data.')

        # 将最后一个元素放到第一个元素位置，然后删除最后一个元素
        self.data[0] = self.data[len(self.data) - 1]
        self.data.pop()
        # 重新计算位置
        self.shift_down(0)
        return e

    def shift_down(self, index: int):
        while self._left_child(index) < len(self.data):
            j = self._left_child(index)
            if j + 1 < len(self.data) and self.data[j + 1] > self.data[j]:
                j = self._right_child(index)

            if self.data[index] >= self.data[j]:
                break
            self.data[index], self.data[j] = self.data[j], self.data[index]
            index = j

File: golang/priority_queue/heap.go
package priority_queue

import "errors"

type Heap struct {
	data []int
}

// 返回堆大小
func (heap Heap) Size() int {
	return len(heap.data)
}

//  返回是否为空
func (heap Heap) IsEmpty() bool {
	return heap.Size() == 0
}

// 返回parent所在的索引位置
func (heap Heap) Parent(index int) int {
	if index == 0 {
		panic("error")
	}
	return (index - 1) / 2
}

// 返回left child所在的索引位置
func (heap Heap) LeftChild(index int) int {
	return index*2 + 1
}

// 返回right child所在的索引位置
func (heap Heap) RightChild(index int) int {
	return (index + 1) * 2
}

// 添加元素
func (heap *Heap) Add(e int) {
	heap.data = append(heap.data, e)
	heap.shiftUp(len(heap.data) - 1)
}

// 上浮操作
func (heap *Heap) shiftUp(index int)  {
	for index > 0 && heap.data[heap.Parent(index)] < heap.data[index] {
		heap.data[index], heap.data[heap.Parent(index)] = heap.data[heap.Parent(index)], heap.data[index]
		index = heap.Parent(index)
	}
}

// 获取最大元素
func (heap *Heap) ExtractMax()(maxEle int, err error){
	if len(heap.data) == 0{
		return 0, errors.New("empty data")
	}
	maxEle = heap.data[0]
	heap.data[0] = heap.data[len(heap.data) - 1]
	// pop
	heap.data = heap.data[:len(heap.data) - 1]
	heap.shiftDown(0)
	return
}

// 下潜操作
func (heap *Heap) shiftDown(index int){
	for heap.LeftChild(index) < len(heap.data){
		j := heap.LeftChild(index)
		if j + 1 < len(heap.data) && heap.data[j+ 1] > heap.data[j]{
			j = heap.RightChild(index)
		}
		if heap.data[index] >= heap.data[j] {
			break
		}
		heap.data[index], heap.data[j] = heap.data[j], heap.data[index]
		index = j
	}
}




